### DISCUSSION TITLE
Can we mutate state in NavigationContainer#onStateChange?

### DISCUSSION THREAD

**Discussion by bogas04** [react-navigation/react-navigation#10142] [Q&A] [answered]
In our app we were trying to limit the max number of routes in the stack to prevent high memory usage on lowend devices. We attempted to solve this problem by listening to state changes and reset it in case total number of routes exceeded an arbitrary number.

```tsx
<NavigationContainer
  onStateChange={(state) => {
    if (state.routes.length > 7) {
      let routes = state.routes;
      routes.shift();
      ScreenRegistry.rootNavigator.dispatch(
        CommonActions.reset({
          ...state,
          index: routes.length - 1,
          routes,
        })
      );
    }
  }}
>
  ...
</NavigationContainer>
```

1. Do you think this is a good way to handle this? If not, how would you tackle this problem?
2. We are seeing this error "Cannot read property 'key' of undefined" [here](https://github.com/react-navigation/react-navigation/blob/26ba019155e2638035a71f755330b52ff281b196/packages/core/src/useNavigationCache.tsx#L144) on certain navigation events. I suspect is due to a race-condition. It might be due to us mutating the state during a navigation event, as the error implies state.index doesn't exist in state.routes.
<img src="https://user-images.githubusercontent.com/6177621/142820946-9bbb4690-8db3-47b6-a75c-c01729898659.png" alt="image of the error" width="200" />

**Comment by satya164** [comment]
No, the library isn't built to deal with mutable state. You also don't need to mutate anything if you're going to dispatch a reset action anyway.

**Reply by bogas04** [reply]
Thanks for quick answer @satya164!

Do you think it's safe to dispatch a reset (without any mutations) inside of onStateChange?

**Comment by satya164** [comment, marked as answer]
Technically it's ok but I can imagine it glitching animations. Though using `react-native-screens` should already improve memory usage. If you must reset on state change, I'd suggest to do it in a `InteractionManager.runAfterInteractions` callback.
